code/game/objects/structures/janicart.dm

```diff
@@ -6,72 +6,138 @@
+	var/obj/structure/mopbucket/mybucket = null
+	var/has_items = FALSE
+	var/dismantled = TRUE
 	var/signs = 0	//maximum capacity hardcoded below
 
 
 /obj/structure/janitorialcart/New()
-	create_reagents(460)
+	//create_reagents(460)
 
 
+/obj/structure/janitorialcart/Destroy()
+	QDEL_NULL(mybag)
+	QDEL_NULL(mymop)
+	QDEL_NULL(myspray)
+	QDEL_NULL(myreplacer)
+	QDEL_NULL(mybucket)
+	return ..()
+
 /obj/structure/janitorialcart/examine(mob/user)
 	if(..(user, 1))
-		user << "[src] \icon[src] contains [reagents.total_volume] unit\s of liquid!"
-	//everything else is visible, so doesn't need to be mentioned
+		if (mybucket)
+			var/contains = mybucket.reagents.total_volume
+			user << "\icon[src] The bucket contains [contains] unit\s of liquid!"
+		else
+			user << "\icon[src] There is no bucket mounted on it!"
 
+/obj/structure/janitorialcart/MouseDrop_T(atom/movable/O as mob|obj, mob/living/user as mob)
+	if (istype(O, /obj/structure/mopbucket) && !mybucket)
+		O.forceMove(src)
+		mybucket = O
+		user << "You mount the [O] on the janicart."
+		update_icon()
+	else
+		..()
 
 /obj/structure/janitorialcart/attackby(obj/item/I, mob/user)
-	if(istype(I, /obj/item/weapon/storage/bag/trash) && !mybag)
-		user.unEquip(I, src)
-		mybag = I
-		update_icon()
-		updateUsrDialog()
-		user << SPAN_NOTICE("You put [I] into [src].")
+	if(istype(I, /obj/item/weapon/mop) || istype(I, /obj/item/weapon/reagent_containers/glass/rag) || istype(I, /obj/item/weapon/soap))
+		if (mybucket)
+			if(I.reagents.total_volume < I.reagents.maximum_volume)
+				if(mybucket.reagents.total_volume < 1)
+					user << "<span class='notice'>[mybucket] is empty!</span>"
+				else
+					mybucket.reagents.trans_to_obj(I, 5)	//
+					user << "<span class='notice'>You wet [I] in [mybucket].</span>"
+					playsound(loc, 'sound/effects/slosh.ogg', 25, 1)
+			else
+				user << "<span class='notice'>[I] can't absorb anymore liquid!</span>"
+		else
+			user << "<span class='notice'>There is no bucket mounted here to dip [I] into!</span>"
+		return 1
 
-	else if(istype(I, /obj/item/weapon/mop))
+	else if (istype(I, /obj/item/weapon/reagent_containers/glass/bucket) && mybucket)
+		I.afterattack(mybucket, usr, 1)
 		return
 
 	else if(istype(I, /obj/item/weapon/reagent_containers/spray) && !myspray)
-		user.unEquip(I, src)
+		user.drop_from_inventory(I,src)
 		myspray = I
 		update_icon()
 		updateUsrDialog()
-		user << SPAN_NOTICE("You put [I] into [src].")
+		user << "<span class='notice'>You put [I] into [src].</span>"
+		return 1
 
 	else if(istype(I, /obj/item/device/lightreplacer) && !myreplacer)
-		user.unEquip(I, src)
+		user.drop_from_inventory(I,src)
 		myreplacer = I
 		update_icon()
 		updateUsrDialog()
-		user << SPAN_NOTICE("You put [I] into [src].")
+		user << "<span class='notice'>You put [I] into [src].</span>"
+		return 1
+
+	else if(istype(I, /obj/item/weapon/storage/bag/trash) && !mybag)
+		user.drop_from_inventory(I,src)
+		mybag = I
+		I.forceMove(src)
```

Comment:
Same here. Also loc passed to `unEquip`/`drop_from_inventory` as second arc so there is no need in `forceMove` call here.